### PR TITLE
ReplicationThreadState: multi-valued logic

### PR DESCRIPTION
**Pull request by shlomi-noach**
Fixes https://github.com/github/orchestrator/issues/763

Until today, replication thread state was a boolean: `true` when running, `false` otherwise. However, a IO thread could also be in a `Connecting` state, which is neither running nor stopped. And this state, as described in #763, can lead to wrong assumptions.

This PR introduces multi-valued logic. `ReplicationthreadState` is introduced, which can be:

- `ReplicationthreadStateStopped`
- `ReplicationthreadStateRunning`
- `ReplicationthreadStateOther`

In this initial commit, most critical existing logic has been converted to use these states as opposed to relying on a boolean value. 

There's more places in the code to modify, in next commits.

cc @ggunson

### CODE REVIEW

**Comment by shlomi-noach** [comment]
Added `ReplicationThreadStateNoThread` state, to strictly identify the case of a master: no replication configured.

**Comment by ggunson** [comment]
Just to clarify the purpose of https://github.com/github/orchestrator/issues/763:

https://github.com/github/orchestrator/issues/763 is not solely about the fact that the IO thread can be in the "Connecting" state, but that the logic of determining the replication status was incomplete.

If true is 

> (SQL_Running=Yes) && (IO_Running=Yes)

Then the opposite of that, based on the distribution rules of negation, is:

> !((SQL_Running=Yes) && (IO_Running=Yes)) = !(SQL_Running=Yes) || !(IO_Running=Yes)

This then leads to all the states that the logic would say replication is stopped:

1. IO_Running=No
   SQL_Running=No 

1. IO_Running=No
   SQL_Running=Yes

1. IO_Running=Yes
   SQL_Running=No

1. IO_Running=Connecting
   SQL_Running=Yes

1. IO_Running=Connecting
   SQL_Running=No

1. (Anything else I might've missed?)

And that was the concern, more than just the extra value that Slave_IO_Running can be. This PR looks like it's dealing with all of that so far; this is just to document the extent of the reasoning for anyone reading this.

File: go/inst/instance.go [hunk 1]
```diff
@@ -255,12 +257,17 @@ func (this *Instance) IsMaster() bool {
 
 // ReplicaRunning returns true when this instance's status is of a replicating replica.
 func (this *Instance) ReplicaRunning() bool {
-	return this.IsReplica() && this.Slave_SQL_Running && this.Slave_IO_Running
+	return this.IsReplica() && this.ReplicationSQLThreadState.IsRunning() && this.ReplicationIOThreadState.IsRunning()
 }
 
 // NoReplicationThreadRunning returns true when neither SQL nor IO threads are running (including the case where isn't even a replica)
```
Comment (ggunson) [inline go/inst/instance.go:263]: How does this return true if the host isn't a replica? `IsStopped()` maps to `ReplicationThreadStateStopped` but doesn't include `ReplicationThreadStateNoThread`? Or am I reading this incorrectly?

Comment (shlomi-noach) [reply go/inst/instance.go:263]: If the host isn't a replica this method returns `false`, which is the expected behavior.

**Pull request merged**